core/crypto/eddsa/blinded25519.go

```diff
@@ -0,0 +1,169 @@
+// There doesn't seem to be another API that lets us initialize 32-byte scalars
+// directly.
+
+// And lastly, note that it would be interesting to come up with a
+// blinding scheme that preserves the "validity" of the derived secret key
+// for use with x25519, perhaps a way to modify `factor` such that
+//            clamp(a)*clamp(factor)
+// ===  clamp(clamp(a)*clamp(factor))
+
+package eddsa
+
+import (
+	"crypto/ed25519"
+	"crypto/sha512"
+	"filippo.io/edwards25519"
+)
+
+type BlindedPrivateKey struct {
+	blinded ed25519.PrivateKey
+}
+
+func (bpk *BlindedPrivateKey) PublicKey() *PublicKey {
+	pub := new(PublicKey)
+	err := pub.FromBytes(bpk.blinded[32:])
+	if err != nil {
+		// This should not be possible:
+		panic(err)
+	}
+	return pub
+}
+
+func (privateKey *BlindedPrivateKey) Sign(message []byte) [64]byte {
+	signature := [64]byte{}
+	// vendored version of ed25519.sign() except it uses the
+	// secret/private scalar directly as expandedSecretKey
+	// instead of deriving from hash each time.
+	// It is a bit unfortunate that this needs to be here...
+	blindedSecretKey := [64]byte{}
+	copy(blindedSecretKey[:], privateKey.blinded[:32])
+	expandedSecretKey := new(edwards25519.Scalar)
+	expandedSecretKey.SetUniformBytes(blindedSecretKey[:])
+
+	// secret_salt <- sha512(private)
+	digest1 := sha512.Sum512(privateKey.blinded[:32])
+
+	// very important that nonce is never repeated for different messages;
+	// ensured here by deriving nonce from a hash of the message.
+	// the second appendix of secret_salt below is not part of the spec:
+	// nonce := sha512(secret_salt[32:64] || message || secret_salt[33:64])
+	h := sha512.New()
+	h.Write(digest1[32:])
+	h.Write(message)
+	h.Write(digest1[33:])
+	messageDigest := h.Sum(nil)
+
+	// messageDigestReduced <- (nonce mod L)
+	mdReduced, _ := new(edwards25519.Scalar).SetUniformBytes(messageDigest[:64])
+
+	// R <- messageDigestReduced * B
+	var encodedR [32]byte
+	copy(encodedR[:], new(edwards25519.Point).ScalarBaseMult(mdReduced).Bytes())
+
+	// hramDigestReduced := sha512(R || public || message) mod L
+	h.Reset()
+	h.Write(encodedR[:])
+	h.Write(privateKey.blinded[32:])
+	h.Write(message)
+	hramDigest := h.Sum(nil)
+	hramDigestReduced, _ := new(edwards25519.Scalar).SetUniformBytes(hramDigest[:])
+	//     s := (hRamDigestReduced*secret + messageDigestReduced) mod L
+	// <-> s := (   (sha512(R || public || message) mod L)*secret
+	//            + (nonce mod L)
+	//          ) mod L
+	s_new := new(edwards25519.Scalar).MultiplyAdd(hramDigestReduced,
+		expandedSecretKey,
+		mdReduced)
+
+	copy(signature[:], encodedR[:])
+	copy(signature[32:], s_new.Bytes())
+	return signature
+}
+
+func hash_factor_inplace(factor *[32]byte) {
+	h := sha512.Sum512_256(factor[:])
+	copy(factor[:], h[:])
+	// needs to be clamped for scalar multiplication,
+	// but Scalar.SetBytesWithClamping takes care of that.
+}
+
+func (secret *PrivateKey) Blind(factor [32]byte) *BlindedPrivateKey {
```

Comment:
it would be better to pass a reference to factor or use the []byte type

Reply:
is the intent to modify the callers copy of factor? I refactored this a little bit and basically Blind(factor[32]byte) is passing a COPY of factor to Blind() and then modifying it. but the caller of Blind()'s copy of factor is not modified.

in my refactored version Blind() is instead passed a byte slice []byte and doesn't use hash_factor_inplace because that then DOES modify the caller of Blind()'s copy of factor (which caused the unit tests to fail, so I think that isn't the expected behavior)

Reply:
answered below re: copying factor.
Hmm, if we use a `[]byte` slice then we open up the possibility for API misuse passing a non-32-byte factor, and then we need a check for that and a corresponding exception/error, and then callers need to handle that. I think that adds unnecessary complications for the users compared to the cost of copying the additional eight bytes.

Reply:
@mixmasala in general we don't want to modify blinding factors... instead we use blinding factors to "bind" a public key... mutate public key or return a new mutated public key.